File: Cap 10/10.8 - Recursividade ou Recorrencia/10.8.2 - Recursividade de Cauda/recauda.lua
--[[
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.8
 * 
 * Programa ..: FUNCAO COM RECURSIVDADE DE CAUDA
--]]

local function FATORIALBASE(N, P)
  if (N == 0) then
    return P
  else
    return FATORIALBASE(N - 1, N * P)
  end
end

local function FATORIAL(N)
  return FATORIALBASE(N, 1)
end

print("CALCULO DE FATORIAL")
print("RECURSIVIDADE DE CAUDA")
print()

io.write("Qual fatorial (1-20): ")
local LIMITE = tonumber(io.read())

print("Fatorial = " .. string.format("%.0f", FATORIAL(LIMITE)))

File: Cap 10/10.8 - Recursividade ou Recorrencia/10.8.2 - Recursividade de Cauda/recauda.pas
{*
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.8
 * 
 * Programa ..: FUNCAO COM RECURSIVDADE DE CAUDA
 * 
 *}

program CALC_FAT_V5;

function FATORIALBASE(N: byte; P: qword): qword;
begin
  if (N = 0) then
    FATORIALBASE := P
  else
    FATORIALBASE := FATORIALBASE(N - 1, N * P);
end;

function FATORIAL(N: byte): qword;
begin
  FATORIAL := FATORIALBASE(N, 1);
end;

var
  LIMITE: byte;

begin

  writeln('CALCULO DE FATORIAL');
  writeln('RECURSIVIDADE DE CAUDA');
  writeln;
  
  write('Qual fatorial (1-20): ');
  readln(LIMITE);
  
  writeln('Fatorial = ', FATORIAL(LIMITE));
  
end.
